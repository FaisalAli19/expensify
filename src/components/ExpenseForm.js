import React, { Component } from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends Component {
	constructor(props) {
		super(props)
		this.state = {
			description: props.expense ? props.expense.description : '',
			amount: props.expense ? (props.expense.amount / 100).toString() : '',
			note: props.expense ? props.expense.note : '',
			createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
			focused: props.expense ? props.expense.focused : false,
			error: ''
		}
	}
	onDescriptionChange = (e) => {
		const description = e.target.value
		this.setState({ description })
	}
	onAmountChange = (e) => {
		const amount = e.target.value
		if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
			this.setState({ amount })
		}
	}
	onNoteChange = (e) => {
		const note = e.target.value
		this.setState({ note })
	}
	onDateChange = (createdAt) => {
		if (createdAt) {
			this.setState({ createdAt })
		}
	}
	onFocusedChange = ({ focused }) => {
		this.setState({ focused })
	}
	onSubmit = (e) => {
		e.preventDefault();
		const { description, amount, createdAt, note } = this.state

		if (!description || !amount) {
			let error = 'Please provide description and amount!'
			//Set error
			this.setState({ error })
		} else {
			this.setState({ error: '' })
			this.props.onSubmit({
				description,
				note,
				amount: parseFloat(amount, 10) * 100,
				createdAt: createdAt.valueOf()
			})
		}
	}
	render() {
		const { description, amount, note, createdAt, focused } = this.state;
		return (
			<form className="form" onSubmit={this.onSubmit}>
				{ this.state.error && <p className="form__error">{this.state.error}</p>}
					<input type="text" 
						className="text-input"
						placeholder="Description" autoFocus 
						value={ description } 
						onChange={this.onDescriptionChange}
					/>
					<input type="text" 
					className="text-input"
					placeholder="Amount" 
					value={ amount }
					onChange={this.onAmountChange}
					/>
					<SingleDatePicker 
						date={createdAt}
						onDateChange={this.onDateChange}
						focused={focused}
						onFocusChange={ this.onFocusedChange }
						numberOfMonths={1}
						isOutsideRange={() => false} //Enable prior date to current date
					/>
					<textarea 
						className="textarea-input"
						placeholder="Add a note for your expense (optional)"
						value={ note }
						onChange={this.onNoteChange}
						></textarea>
					<div>
						<button className="buttons">Add Expense</button>
					</div>
			</form>
		)
	}
}

export default ExpenseForm