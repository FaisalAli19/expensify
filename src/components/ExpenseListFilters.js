import React, { Component } from 'react';
import { connect } from 'react-redux'
import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';
import uuid from 'uuid'

import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters'

export class ExpenseListFilters extends Component {
	state = {
		calendarFocused: null
	}
	onDatesChange = ({ startDate, endDate }) => {
		const { setStartDate, setEndDate } = this.props
		setStartDate(startDate)
		setEndDate(endDate)
	}
	onFocusChange = (calendarFocused) => {
		this.setState(() => ({ calendarFocused }))
	}
	onTextChange = (e) => {
		this.props.setTextFilter(e.target.value)
	}
	onFilterByChange = (e) => {
		const { sortByDate, sortByAmount } = this.props
		e.target.value === 'date' ? sortByDate() : sortByAmount()
	}
	render() {
		const { text, sortBy, startDate, endDate } = this.props.filters;
		return (
			<div className="content-container">
				<div className="input-group">
					<div className="input-group__item">
						<input className="text-input" type="text" value={text} 
							onChange={this.onTextChange}
							placeholder="Search expenses"
						/>
					</div>
					<div className="input-group__item">
						<select className="select-input" value={sortBy} onChange={this.onFilterByChange}>
							<option value="date">Date</option>
							<option value="amount">Amount</option>
						</select>
					</div>
					<div className="input-group__item">
						<DateRangePicker 
							startDate={startDate}
							startDateId={'1'}
							endDate={endDate}
							endDateId={'2'}
							onDatesChange={this.onDatesChange}
							focusedInput={this.state.calendarFocused}
							onFocusChange={this.onFocusChange}
							numberOfMonths={1}
							isOutsideRange={() => false}
							showClearDates={true}
						/>
					</div>
				</div>
			</div>
		)
	}
}

const mapStateToProps = state => {
	return {
		filters: state.filters
	};
}

const mapDispatchToProps = (dispatch) => ({ 
	setTextFilter: (text) => dispatch(setTextFilter(text)), 
	sortByDate: () => dispatch(sortByDate()), 
	sortByAmount: () => dispatch(sortByAmount()), 
	setStartDate: (date) => dispatch(setStartDate(date)), 
	setEndDate : (date) => dispatch(setEndDate(date))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)