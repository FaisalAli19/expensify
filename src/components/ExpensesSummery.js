import React from 'react'
import { connect } from 'react-redux';
import numeral from 'numeral';

import selectExpenses from '../selectors/expenses'
import selectedExpensesTotal from '../selectors/expense-total';


export const ExpensesSummery = ({ expenses }) => {
	const expenseWord = expenses.length === 1 ? 'expense' : 'expenses'
	return (
		<p>
			Viewing { expenses.length } { expenseWord } totalling { 
				numeral(selectedExpensesTotal(expenses) / 100).format('$0,0')
			}
		</p>
	)
};

const mapStateToProps = state => {
	return {
		expenses: selectExpenses(state.expenses, state.filters)
	}
}

export default connect(mapStateToProps)(ExpensesSummery);