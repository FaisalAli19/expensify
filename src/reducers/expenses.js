// Expenses Reducers

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
	switch (action.type) {
		case 'ADD_EXPENSE':
			return [ ...state, action.expense ]
		case 'REMOVE_EXPENSE':
			return state.filter(({ id }) => id !== action.id )
		case 'EDIT_EXPENSE':
			return state.map((exp) => {
				if (exp.id === action.id) {
					return { ...exp, ...action.updates };
				} else {
					return exp
				}
			})
		default:
			return state;
	}
};