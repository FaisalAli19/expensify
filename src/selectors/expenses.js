import moment from 'moment'

// Get Visible expenses
export default (expenses, { text, sortBy, startDate, endDate }) => {
	return expenses.filter(exp => {
		const createdAtMoment = moment(exp.createdAt)
		const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true
		const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
		const textMatch = exp.description.toLowerCase().includes(text.toLowerCase());

		return startDateMatch && endDateMatch && textMatch;
	}).sort((a, b) => {
		if (sortBy === 'amount') {
			return a.amount < b.amount ? 1 : -1;
		} else if (sortBy === 'date') {
			return a.createdAt < b.createdAt ? 1 : -1;
		}
	});
}