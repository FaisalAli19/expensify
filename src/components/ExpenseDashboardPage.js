import React from 'react';

import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummery from './ExpensesSummery'

const ExpenseDashboardPage = (props) => {
	return (
		<div>
			<ExpensesSummery />
			<ExpenseListFilters />
			<ExpenseList />
  	</div>
	)
};

const mapStateToProps = state => {
	return {
		expenses: selectExpenses(state.expenses, state.filters)
	}
}

export default ExpenseDashboardPage;
