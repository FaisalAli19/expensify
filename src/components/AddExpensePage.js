import React, { Component } from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { addExpense } from '../actions/expenses'

export class AddExpensePage extends Component {
	onSubmit = (expense) => {
		const { addExpense, history } = this.props
		addExpense(expense)
		history.push('/')
	}
	render() {
		return (
			<div>
				<h2>Add Expense</h2>
				<ExpenseForm onSubmit={this.onSubmit} />
  		</div>
		)
	}
}

const mapDispatchToProps = (dispatch) => ({
	addExpense: (expense) => dispatch(addExpense(expense))
})


export default connect(null, mapDispatchToProps)(AddExpensePage);
