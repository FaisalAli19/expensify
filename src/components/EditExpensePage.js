import React, { Component } from 'react'
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { editExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends Component {
	onSubmitForm = (updatedExpense) => {
		const { editExpense, history, expense } = this.props
		editExpense(expense.id, updatedExpense)
		history.push('/')
	}
	onRemove = () => {
		const { startRemoveExpense, expense, history } = this.props
		startRemoveExpense(expense.id)
		history.push('/')
	}
	render() {
		return (
			<div>
				<ExpenseForm  expense={this.props.expense} onSubmit={this.onSubmitForm} />
				<button onClick={this.onRemove}>Remove</button>
  		</div>
		)
	}
}

const mapStateToProps = (state, props) => ({
	expense: state.expenses.find(exp => exp.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch) => ({ 
		editExpense: (id, expense) => dispatch(editExpense(id, expense)), 
		startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
})


export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
