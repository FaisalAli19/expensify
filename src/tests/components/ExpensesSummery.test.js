import React from 'react'
import { shallow } from 'enzyme';

import { ExpensesSummery } from '../../components/ExpensesSummery';
import expenses from '../fixtures/expenses'

test('should render correctly with no data', () => {
	const wrapper = shallow(<ExpensesSummery expenses={[]} />)
	expect(wrapper).toMatchSnapshot();
});

test('should render correctly with expense data', () => {
	const wrapper = shallow(<ExpensesSummery expenses={[expenses[0]]} />)
	expect(wrapper).toMatchSnapshot();
});

test('should render correctly with expenses data', () => {
	const wrapper = shallow(<ExpensesSummery expenses={expenses} />)
	expect(wrapper).toMatchSnapshot();
});