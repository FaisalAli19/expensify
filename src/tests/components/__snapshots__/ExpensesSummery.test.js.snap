// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly with expense data 1`] = `
<p>
  Viewing 
  1
   
  expense
   totalling 
  $2
</p>
`;

exports[`should render correctly with expenses data 1`] = `
<p>
  Viewing 
  3
   
  expenses
   totalling 
  $67
</p>
`;

exports[`should render correctly with no data 1`] = `
<p>
  Viewing 
  0
   
  expenses
   totalling 
  $0
</p>
`;
